### PR TITLE
using chrono and gtk::timeout_add_seconds to build a basic clock

### PR DESCRIPTION
**Pull request by n3phtys**
How to make a minimal clock with gtk-rs + chrono only.

I had to use this code in a personal project, and IMO it could save someone new 2 minutes to have this example.

The gtk code is adapted from basic.rs but with a Label instead of a Button.

### CODE REVIEW

**Comment by GuillaumeGomez** [comment]
Awesome, thanks! (You even opened the PR on the good target branch haha)

Just a few things though:

 * Please make imports by blocks, such as:

  ```rust
  extern crate a;
  extern crate b;

  use a::x;
  use b::y;
  ```

* Please put the `main` function at the end.
* Please use `Application` instead of `gtk::main`.

**Comment by GuillaumeGomez** [comment]
Just one last nit: please add a whitespace after `//`.

Does it seem good for you @EPashkin and @sdroege?

File: src/bin/basic.rs [hunk 1]
```diff
@@ -51,9 +51,8 @@ fn build_ui(application: &gtk::Application) {
 }
 
 fn main() {
-    let application = gtk::Application::new("com.github.basic",
-                                            gio::ApplicationFlags::empty())
-                                       .expect("Initialization failed...");
+    let application = gtk::Application::new("com.github.basic", gio::ApplicationFlags::empty())
```
Comment (GuillaumeGomez) [inline src/bin/basic.rs]: Ugly.

File: src/bin/builder_basics.rs [hunk 2]
```diff
@@ -40,10 +40,12 @@ mod example {
...
         let builder = Builder::new_from_string(glade_src);
 
-        let window: ApplicationWindow = builder.get_object("window1").expect("Couldn't get window1");
+        let window: ApplicationWindow =
+            builder.get_object("window1").expect("Couldn't get window1");
         let bigbutton: Button = builder.get_object("button1").expect("Couldn't get button1");
-        let dialog: MessageDialog = builder.get_object("messagedialog1")
-                                           .expect("Couldn't get messagedialog1");
+        let dialog: MessageDialog = builder
+            .get_object("messagedialog1")
```
Comment (GuillaumeGomez) [inline src/bin/builder_basics.rs]: ugly

File: src/bin/builder_basics.rs [hunk 3]
```diff
@@ -40,10 +40,12 @@ mod example {
         let glade_src = include_str!("builder_basics.glade");
         let builder = Builder::new_from_string(glade_src);
 
-        let window: ApplicationWindow = builder.get_object("window1").expect("Couldn't get window1");
+        let window: ApplicationWindow =
+            builder.get_object("window1").expect("Couldn't get window1");
```
Comment (GuillaumeGomez) [inline src/bin/builder_basics.rs]: useless

File: src/bin/builder_basics.rs [hunk 4]
```diff
@@ -60,9 +62,9 @@ mod example {
     }
 
     pub fn main() {
-        let application = gtk::Application::new("com.github.builder_basics",
-                                                gio::ApplicationFlags::empty())
-                                           .expect("Initialization failed...");
+        let application =
+            gtk::Application::new("com.github.builder_basics", gio::ApplicationFlags::empty())
```
Comment (GuillaumeGomez) [inline src/bin/builder_basics.rs]: ugly

**Comment by GuillaumeGomez** [comment]
Please don't run rustfmt on the examples (so remove your last commit). It's ugly and not readable.

**Comment by n3phtys** [comment]
This should definitely be remarked in the CONTRIBUTING.md

**Comment by GuillaumeGomez** [comment]
Revert isn't remove. Please [squash](https://blog.guillaume-gomez.fr/Git-tips/1/2).

**Comment by GuillaumeGomez** [comment]
Well, why would you want to run rustfmt on the codebase? :-/

**Comment by n3phtys** [comment]
Both commits are now removed from history.

rustfmt is just one example. I think it's pretty general to have standardized code styles that can be checked automatically. rustfmt is just one standard.

**Comment by sdroege** [comment]
> Well, why would you want to run rustfmt on the codebase? :-/

Having one enforced code style definitely makes sense, for consistency and simplicity. And rustfmt nowadays (rustfmt-nightly!) by default outputs something nice (and can be configured if needed), I think it's a good choice. Should be done as a separate commit though.

**Comment by GuillaumeGomez** [comment]
It's up to debate. If the generated code looks nice, I have no issue with it. However, as long as rustfmt does this:

```rust
something.yolo()
    .a()
   .b();
```

Instead of:

```rust
something.yolo()
         .a()
         .b();
```

I'm strongly against it.

**Comment by n3phtys** [comment]
This behavior can be changed via: [chain_indent](https://github.com/rust-lang-nursery/rustfmt/blob/master/Configurations.md#chain_indent)

**Comment by GuillaumeGomez** [comment]
That's a lot of efforts for not much gains. But whatever, it's the debate in here. If you *really* want to add it, just open an issue and we'll discuss about it.

File: src/bin/clock.rs [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+    window.set_default_size(260, 40);
+
+    window.connect_delete_event(clone!(window => move |_, _| {
+        window.destroy();
+        Inhibit(false)
+    }));
+
+    let time = current_time();
+    let label = gtk::Label::new(None);
+    label.set_text(&time);
```
Comment (EPashkin) [inline src/bin/clock.rs]: Maybe just `let label = gtk::Label::new(&*time);`

**Comment by EPashkin** [comment]
@n3phtys Thanks

@GuillaumeGomez 👍 after squash.

**Comment by GuillaumeGomez** [comment]
Hi. Please [squash](https://blog.guillaume-gomez.fr/Git-tips/1/2) your commits so we can merge this.

**Comment by EPashkin** [comment]
@n3phtys sorry, forgot about this PR
@GuillaumeGomez Please restart travis or just merge.

**Comment by GuillaumeGomez** [comment]
Oups, thanks!

**Pull request merged**